File: src/forge/agents/verification.py
import json
from pathlib import Path
from forge.agents.base import BaseAgent, AgentResult, _extract_json
from forge.router import ModelTier

SYSTEM = """You are a QA engineer verifying that a project builds and its tests pass.

You have tools available:
- bash_exec: run build commands, test suites, linters
- read_file: read files to understand failures
- write_file: apply quick fixes for obvious issues (wrong import path, missing config)
- list_dir: list directory contents

Workflow:
1. Use list_dir to understand the project structure
2. Run the build (e.g. `npm run build` or `python -m pytest`) with bash_exec
3. If it fails: read the relevant source files, understand the error, apply a targeted fix
4. Re-run to confirm the fix worked
5. Run the test suite after a successful build
6. When satisfied (build passes, tests pass or are acceptably skipped), output a JSON report:

{
  "passed": ["Build succeeded", "All 5 tests passed"],
  "failed": [],
  "errors": []
}

If the build failed after your best attempts:
{
  "passed": [],
  "failed": ["Build failed: <reason>"],
  "errors": ["<raw error snippet>"]
}

Output ONLY the JSON report as your final message. Do not wrap it in markdown."""


class VerificationAgent(BaseAgent):
    tier = ModelTier.REASONING

    async def run(
        self,
        workspace: Path,
        architecture: str,
        spec: str,
    ) -> AgentResult:
        messages: list[dict] = [
            {"role": "system", "content": SYSTEM},
            {"role": "user", "content": (
                f"Architecture:\n{architecture}\n\n"
                f"Spec:\n{spec}\n\n"
                f"Workspace root: {workspace}"
            )},
        ]
        response = await self._run_agentic_loop(messages=messages, workspace=workspace)

        try:
            report: dict = json.loads(_extract_json(response))
        except json.JSONDecodeError:
            report = {
                "passed": [],
                "failed": ["Verification agent returned malformed report"],
                "errors": [response[:300]],
            }

        success = len(report.get("failed", [])) == 0 and len(report.get("errors", [])) == 0
        return AgentResult(
            success=success,
            output=json.dumps(report),
            error=None if success else "verification_failed",
        )

File: src/agents/verification.ts
import { BaseAgent, AgentResult, type AgentRunOptions } from "./base.js";
import { ModelTier } from "../router.js";

const SYSTEM = `You are a QA engineer verifying that a project builds and its tests pass.

You have tools available:
- bash_exec: run build commands, test suites, linters
- read_file: read files to understand failures
- write_file: apply quick fixes for obvious issues
- list_dir: list directory contents

Workflow:
1. Use list_dir to understand the project structure
2. Run the build (e.g. \`npm run build\` or \`python -m pytest\`) with bash_exec
3. If it fails: read the relevant source files, understand the error, apply a targeted fix
4. Re-run to confirm the fix worked
5. Run the test suite after a successful build
6. When satisfied, output a JSON report:

{
  "passed": ["Build succeeded", "All 5 tests passed"],
  "failed": [],
  "errors": []
}

Output ONLY the JSON report as your final message. Do not wrap it in markdown.`;

function toStringList(value: unknown): string[] {
  if (value === undefined || value === null) return [];
  const items = Array.isArray(value) ? value : [value];
  return items
    .map((item) => (typeof item === "string" ? item : JSON.stringify(item)))
    .filter((item) => item.trim().length > 0);
}

export class VerificationAgent extends BaseAgent {
  protected tier = ModelTier.REASONING;

  async run(args: Record<string, unknown>): Promise<AgentResult> {
    const workspace = String(args["workspace"] ?? "");
    const messages: any[] = [
      { role: "system", content: SYSTEM },
      { role: "user", content: `Architecture:\n${args["architecture"]}\n\nSpec:\n${args["spec"]}\n\nWorkspace root: ${workspace}` },
    ];
    const opts: AgentRunOptions = { skillContext: args["skillContext"] as AgentRunOptions["skillContext"] };
    const response = await this.runAgenticLoop(messages, workspace, undefined, opts);
    let report: Record<string, string[]>;
    try {
      const parsed = JSON.parse(this.extractJson(response));
      if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) throw new Error("not an object");
      // models sometimes omit keys or return a bare string; coerce to string lists
      report = {
        passed: toStringList(parsed["passed"]),
        failed: toStringList(parsed["failed"]),
        errors: toStringList(parsed["errors"]),
      };
    } catch {
      report = { passed: [], failed: ["Verification agent returned malformed report"], errors: [response.slice(0, 300)] };
    }
    const success = report["failed"].length === 0 && report["errors"].length === 0;
    return { success, output: JSON.stringify(report), error: success ? undefined : "verification_failed" };
  }
}
